kernel/net/net_tcp.c

```diff
@@ -1890,6 +1890,91 @@ static int net_tcp_setsockopt(net_socket_t *hnd, int level, int option_name,
     return 0;
 }
 
+static int net_tcp_getsockname(net_socket_t *hnd, struct sockaddr *name, socklen_t *name_len) {
+    int tmp;
+    struct tcp_sock *sock;
+    struct sockaddr_in realaddr;
+    struct sockaddr_in6 realaddr6;
+
+    if(!name || !name_len) {
+        errno = EFAULT;
+        return -1;
+    }
+
+    if(irq_inside_int()) {
+        if(rwsem_read_trylock(&tcp_sem)) {
+            errno = EWOULDBLOCK;
+            return -1;
+        }
+    }
+    else {
+        rwsem_read_lock(&tcp_sem);
+    }
+
+    if(!(sock = (struct tcp_sock *)hnd->data)) {
+        rwsem_read_unlock(&tcp_sem);
+        errno = EBADF;
+        return -1;
+    }
+
+    if(irq_inside_int()) {
+        if(mutex_trylock(&sock->mutex)) {
+            rwsem_read_unlock(&tcp_sem);
+            errno = EWOULDBLOCK;
+            return -1;
+        }
+    }
+    else {
+        mutex_lock(&sock->mutex);
+    }
+
+    if(sock->domain == AF_INET) {
+        memset(&realaddr, 0, sizeof(struct sockaddr_in));
+        realaddr.sin_family = AF_INET;
+        realaddr.sin_addr.s_addr =
+            sock->remote_addr.sin6_addr.__s6_addr.__s6_addr32[3];
```

Comment:
I'm pretty sure this should be reading the local address, since the function is defined to read the locally bound address, not the connected to address. The same applies below in the `AF_INET6` section as well as in the net_udp.c file.